### PR TITLE
feat(worker-loop): WorkerLoop as SelectorEventLoop on the Tokio reactor

### PR DESCRIPTION
**Pull request by FarhanAliRaza**
## Summary
- \`WorkerLoop\` is now a real \`asyncio.SelectorEventLoop\` subclass; the four fd primitives register descriptors with the Tokio reactor directly (\`src/worker_fd.rs\`, Unix).
- Removes the second-loop delegation adapters and \`DJANGO_BOLT_LOOP_CROSSING_TIMEOUT\`; stdlib transports/TLS/DNS/pipes/subprocesses work without cross-loop hops.
- Level-triggered semantics preserved over Tokio's edge-triggered readiness (zero-timeout \`poll(2)\` re-check; \`/t-fd-level-triggered\` probe).
- Updates CLAUDE.md, dispatch probes, and worker-loop tests; ignores \`BLOG_RESEARCH.md\`.

## Test plan
- [ ] \`just build\`
- [ ] \`pytest python/tests/test_worker_loop.py python/tests/integration/test_async_dispatch_server_integration.py -s -vv\`

## Hot request path impact

Yes. The PR changes `WorkerLoop`, which handles request I/O.

Each readiness event now performs:

- Dispatch of the `Ready` command.
- Execution of the registered callback.
- A zero-timeout `poll(2)` check.
- Readiness acknowledgement and watcher re-arm.

This work is required to integrate file descriptors directly with Tokio and preserve asyncio level-triggered behavior. It replaces cross-loop delegation and adapter overhead.

Watcher registration and removal occur only when watchers change. The PR adds no unrelated request processing.

The main added cost is one `poll(2)` check and re-arm message per readiness event. This cost is required for correctness. The PR provides no benchmark data to quantify the impact.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,6 @@ package-lock.json
 yarn.lock
 pnpm-lock.yaml
 .claude/settings.local.json
+docs/src/llms-full.txt
+examples/
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply .gitignore]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: bench/js/compare.sh [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+here="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+repo="$(cd "$here/../.." && pwd)"
+results="$(mktemp)"
+SERVER_PGIDS=()
+
+command -v bombardier >/dev/null || {
+    echo "bombardier not found: go install github.com/codesenberg/bombardier@latest" >&2
+    exit 1
+}
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply bench/js/compare.sh]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: bench/js/compare.sh [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+cleanup() {
+    stop_servers 2>/dev/null || true
+    rm -f "$results"
+}
+trap cleanup EXIT INT TERM
+
+# Launch in its own process group so multi-process runtimes (runbolt workers,
+# node cluster children, bun reusePort siblings) all die with one signal.
+spawn() {
+    setsid "$@" >>"/tmp/bench-hono-compare.log" 2>&1 &
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply bench/js/compare.sh]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: bench/js/compare.sh [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    esac
+    wait_ready "$1"
+}
+
+run_endpoint() {
+    local name="$1" path="$2"
+    local url="http://$HOST:$PORT$path"
+    local bytes
+    bytes="$(curl -fsS "$url" | wc -c | tr -d ' ')"
+
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply bench/js/compare.sh]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: bench/js/server-node-hono.mjs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+if (processes > 1 && cluster.isPrimary) {
+  console.log(
+    `hono/node listening on http://${host}:${port} (${processes} processes, ${availableParallelism()} cores)`,
+  );
+  cluster.schedulingPolicy = cluster.SCHED_NONE; // let the OS balance, like SO_REUSEPORT
+  for (let i = 0; i < processes; i += 1) cluster.fork();
+  cluster.on("exit", (worker, code, signal) => {
+    console.error(`worker ${worker.process.pid} exited (${signal || code})`);
+  });
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply bench/js/server-node-hono.mjs]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+```bash
+# 1. Fork and clone
+git clone https://github.com/<your-username>/django-bolt.git
+cd django-bolt
+
+# 2. Install Python dependencies
+uv sync
+
+# 3. Build the Rust extension (required after any Rust change)
+just build            # or: maturin develop --release
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply CONTRIBUTING.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: docs/build_llms_full.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+def main() -> None:
+    pages = list(ORDER)
+    for group in ("topics", "ref"):
+        pages += sorted(p.relative_to(SRC).as_posix() for p in (SRC / group).glob("*.md"))
+
+    parts = [(SRC / "llms.txt").read_text(), "\n\n---\n\n# Full documentation\n"]
+    for rel in pages:
+        text = FRONTMATTER.sub("", (SRC / rel).read_text())
+        parts.append(f"\n\n<!-- source: {page_url(rel)} -->\n\n{text.strip()}\n")
+    (SRC / "llms-full.txt").write_text("".join(parts))
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply docs/build_llms_full.py]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: docs/src/architecture.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
+---
+icon: lucide/layers
+description: How Django-Bolt works — a Rust (Actix Web + Tokio) HTTP server calls Python handlers through PyO3, with routing, auth, guards, CORS, rate limiting and compression running in Rust without the GIL. No WSGI, no ASGI server.
+---
+
+# How Django-Bolt works
+
+Django-Bolt is **not** a WSGI or ASGI application and does not run behind gunicorn or uvicorn. It is a Rust HTTP server that embeds the Python interpreter and calls your Django handlers directly. This page explains the request path and why it is fast.
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply docs/src/architecture.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: docs/src/architecture.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+## The request path
+
+```
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply docs/src/architecture.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: docs/src/benchmarks.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+## Reproduce
+
+```bash
+git clone https://github.com/dj-bolt/django-bolt.git && cd django-bolt
+uv sync && just build
+go install github.com/codesenberg/bombardier@latest
+
+just save-bench                    # full suite → bench/BENCHMARK_BASELINE.md
+just save-bench 127.0.0.1 8001 100 100000 8   # host port C N processes
+just bench-js                      # Bolt vs Hono vs Elysia
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply docs/src/benchmarks.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: justfile [hunk 11]
```diff
@@ -185,6 +185,17 @@ build-profiling:
...
+# Cross-framework comparison: Django-Bolt vs Hono (node + bun) vs Elysia (bun)
+# Runtimes: `just bench-js` (all) or RUNTIMES="bolt elysia-bun" just bench-js
+bench-js host=host port=port c=c n=n p="1":
+    #!/usr/bin/env bash
+    set -euo pipefail
+    if [ ! -d bench/js/node_modules ]; then
+        echo "Installing JS benchmark dependencies..."
+        (cd bench/js && { command -v bun >/dev/null && bun install || npm install; })
+    fi
+    PROCESSES={{p}} C={{c}} N={{n}} HOST={{host}} PORT={{port}} ./bench/js/compare.sh
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply justfile]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: justfile [hunk 12]
```diff
@@ -262,8 +273,8 @@ docs: docs-serve
...
 # Serve documentation locally
 docs-serve:
-    cd docs && uv run zensical serve -a localhost:8080
+    cd docs && uv run python build_llms_full.py && uv run zensical serve -a localhost:8080
 
-# Build documentation
+# Build documentation (also regenerates llms-full.txt for AI crawlers)
 docs-build:
-    cd docs && uv run zensical build --clean
+    cd docs && uv run python build_llms_full.py && uv run zensical build --clean
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply justfile]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: README.md [hunk 13]
```diff
@@ -1,235 +1,378 @@
...
-[![Ask DeepWiki](https://deepwiki.com/badge.svg)](https://deepwiki.com/FarhanAliRaza/django-bolt)
-[![Discord](https://img.shields.io/discord/1513537500000292894?logo=discord&logoColor=white&label=Discord&color=5865F2)](https://discord.gg/4xErptXK82)
-[![Sponsor](https://img.shields.io/badge/Sponsor-Django%20Bolt-pink)](https://opencollective.com/django-bolt)
+---
+
+**Django-Bolt** is the fastest Python web framework: **300k+ requests/second** on a single 12-core desktop (8 processes, C=100, loopback), ahead of FastAPI and Robyn, and even of Bun-based JavaScript frameworks (Elysia, Hono) on JSON payloads. It is a fully typed API framework for Django. It serves your endpoints from a Rust HTTP server ([Actix Web](https://actix.rs/) + [Tokio](https://tokio.rs/)), bridges to your Python handlers with [PyO3](https://pyo3.rs/), and serializes with [msgspec](https://jcristharif.com/msgspec/) — while everything you love about Django (ORM, Admin, auth, middleware, signals, third-party apps) keeps working out of the box.
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply README.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: README.md [hunk 14]
```diff
@@ -1,235 +1,378 @@
...
+- **Actix Web + Tokio** handle HTTP parsing and responses; **matchit** routes with zero-copy path matching.
+- **Auth, guards, CORS, rate limiting, compression** run in Rust — no GIL, no Python per-request overhead.
+- **msgspec** serialization is 5–10× faster than the standard library; response bodies cross to Rust zero-copy.
+- **Sync-dispatch bypass:** handlers that don't actually await are detected at registration and skip the async bridge entirely.
+- **Registration-time precomputation:** parameter extraction, dependency graphs, and middleware are compiled once, reused forever.
+
+## 🏗️ How it works
 
-### Backers
+```
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply README.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: README.md [hunk 15]
```diff
@@ -1,235 +1,378 @@
...
+- ❓ [FAQ](https://bolt.farhana.li/faq/) · [Comparison vs Ninja / DRF / FastAPI / Litestar](https://bolt.farhana.li/comparison/) · [How it works](https://bolt.farhana.li/architecture/)
 
-- **[Litestar](https://github.com/litestar-org/litestar)** - Our OpenAPI plugin system is adapted from Litestar's excellent architecture. Many architectural decisions around middleware, guards, and route handling were inspired by Litestar's design philosophy.
+## 💖 Sponsors
 
-- **[Robyn](https://github.com/sparckles/Robyn)** - Robyn's Rust-Python integration patterns and performance-first approach influenced our decision to use PyO3 and showed us the potential of Rust-powered Python web frameworks.
+Support Django-Bolt's development by [becoming a sponsor](https://opencollective.com/django-bolt). Your logo will appear here with a link to your website.
 
-### Additional Credits
+<a href="https://opencollective.com/django-bolt/tiers/sponsor/0/website" target="_blank"><img src="https://opencollective.com/django-bolt/tiers/sponsor/0/avatar.svg" /></a>
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply README.md]: Outdated — this file is not part of PR #286 (the PR now contains only the WorkerLoop selector/fd change: src/worker_fd.rs, src/worker_loop.rs, _worker_loop.py and its tests). Not addressed here.

File: src/worker_fd.rs [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        // Wait until the callback ran before re-arming: firing again while it
+        // is still queued would run it twice for one readiness event.
+        if ack_rx.await.is_err() {
+            return;
+        }
+        if !still_ready(raw, direction) {
+            guard.clear_ready();
+        }
+    }
+}
```
Comment (FarhanAliRaza) [inline src/worker_fd.rs:210]: Valid — fixed in 45ca0c09. `run_worker_handle` now returns False for a cancelled handle; the Ready ack carries that bool and `watch` exits on it (mirrors the stdlib selector loop dropping cancelled readers). Regression test: `/t-fd-cancelled-handle` probe asserted in `test_dispatch_probes_worker_loop_default` via new `_core.worker_fd_watcher_count()`; verified failing before the fix.

File: python/tests/integration/test_async_dispatch_server_integration.py [hunk 17]
```diff
@@ -59,37 +59,7 @@ def test_dispatch_probes_in_process():
...
-                # suite; the probe wedges the selector loop for 4 seconds.
-                "DJANGO_BOLT_LOOP_CROSSING_TIMEOUT": "2",
-            }
-        ) as server:
-            assert server.get("/t-crossing-timeout").json() == {"timed_out": True}
-    finally:
-        echo_server.shutdown()
-        echo_server.server_close()
-        echo_thread.join(timeout=2)
+        assert self.rfile.readline() == b"ping\n"
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/tests/integration/test_async_dispatch_server_integration.py]: Valid — fixed in 45ca0c09. `_EchoHandler` records the received lines and the test asserts `_EchoHandler.received == [b"ping\n", b"ping\n"]` on the main thread after the /t-socket and /t-buffered-protocol probes.

File: python/tests/integration/apps/dispatch_probes.py [hunk 18]
```diff
@@ -332,6 +313,59 @@ def readable():
...
+    try:
+        loop.add_reader(read_fd, readable)
+        os.write(write_fd, b"abcd")
+        await asyncio.wait_for(done, 1)
+        return {"chunks": [c.decode() for c in chunks]}
+    finally:
+        loop.remove_reader(read_fd)
+        os.close(read_fd)
+        os.close(write_fd)
+
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/tests/integration/apps/dispatch_probes.py:340]: Declined. The watcher never reads through the AsyncFd — it only uses it for readiness — and the callback runs only after the pump delivered a readiness event: either from Tokio (kernel-reported) or after `still_ready`'s zero-timeout `poll(2)` returned >0. With timeout 0 there is no EINTR window, and EFAULT/EINVAL cannot occur for a one-entry pollfd, so the `rc != 0` error branch is not reachable in practice. In these probes the reads are also sized to bytes already written (`os.read(fd, 1)` after writing 4 bytes; `recv(16)` after the peer sends). Setting the descriptors non-blocking would not change what the probes verify.

File: python/tests/integration/apps/dispatch_probes.py [hunk 19]
```diff
@@ -457,15 +491,12 @@ def resume_writing(self):
...
-        selector_protocol = raw_transport.get_protocol()
-        selector_loop = getattr(loop, "_selector_loop", loop)
-        selector_loop.call_soon_threadsafe(selector_protocol.pause_writing)
+        # trigger the flow-control callbacks the way the transport does:
+        # scheduled on the loop against the protocol the transport owns.
+        protocol = transport.get_protocol()
+        loop.call_soon_threadsafe(protocol.pause_writing)
         await asyncio.wait_for(paused, 1)
-        selector_loop.call_soon_threadsafe(selector_protocol.resume_writing)
+        loop.call_soon_threadsafe(protocol.resume_writing)
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/tests/integration/apps/dispatch_probes.py:523]: Pre-existing on master (probe and its direct pause/resume calls predate this PR; only the comment wording changed here), so out of scope. Note the suggested approach does not work either: `_SelectorDatagramTransport.sendto` only buffers (and thus calls `_maybe_pause_protocol`) when `sock.sendto` raises BlockingIOError, so `set_write_buffer_limits` alone never triggers pause on a UDP socket whose kernel buffer is not full. Will revisit the probe separately.

File: python/tests/integration/apps/dispatch_probes.py [hunk 20]
```diff
@@ -332,6 +314,80 @@ def readable():
...
+        before = _core.worker_fd_watcher_count()
+        handle.cancel()
+        os.write(write_fd, b"x")
+        await asyncio.sleep(0.05)
+        return {"watcher_stopped": _core.worker_fd_watcher_count() == before - 1, "calls": len(calls)}
+    finally:
+        loop.remove_reader(read_fd)
+        os.close(read_fd)
+        os.close(write_fd)
+
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/tests/integration/apps/dispatch_probes.py:363]: Valid — fixed in 625fa9d7. The probe now polls `worker_fd_watcher_count()` with a 2s deadline; it still fails if the watcher never stops.

File: python/tests/integration/test_async_dispatch_server_integration.py [hunk 21]
```diff
@@ -57,39 +57,13 @@ def test_dispatch_probes_in_process():
...
-                # suite; the probe wedges the selector loop for 4 seconds.
-                "DJANGO_BOLT_LOOP_CROSSING_TIMEOUT": "2",
-            }
-        ) as server:
-            assert server.get("/t-crossing-timeout").json() == {"timed_out": True}
-    finally:
-        echo_server.shutdown()
-        echo_server.server_close()
-        echo_thread.join(timeout=2)
+        self.received.append(self.rfile.readline())
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/tests/integration/test_async_dispatch_server_integration.py]: Fixed in 625fa9d7 — annotated as `ClassVar`.

File: python/tests/integration/test_async_dispatch_server_integration.py [hunk 22]
```diff
@@ -201,6 +175,7 @@ def log_message(self, format, *args):
             assert response.status_code == 200
             assert response.json() == {"greeting": "hello"}
             assert server.get("/t-buffered-protocol").json() == {"greeting": "hello"}
+            assert _EchoHandler.received == [b"ping\n", b"ping\n"]
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/tests/integration/test_async_dispatch_server_integration.py]: Valid — fixed in 625fa9d7. The handler appends under a `threading.Condition` and the test uses `_EchoHandler.wait_for_lines(2)` (deadline-bounded); the list is cleared before the echo server starts.

File: src/worker_fd.rs [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+        // drops cancelled readers) instead of requeueing while the
+        // descriptor stays ready.
+        if !matches!(ack_rx.await, Ok(true)) {
+            return;
+        }
+        if !still_ready(raw, direction) {
+            guard.clear_ready();
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply src/worker_fd.rs:210]: Documented in 625fa9d7 (`src/worker_fd.rs` module docs): a watched descriptor must be removed before it is closed because the watcher owns a `dup(2)`. asyncio's transports and `sock_*` helpers already do this; a close-without-remove of a raw fd is user error under the stdlib loop too (the selector registration is left dangling there). Not adding a dedicated test: the observable effect is a retained duplicate descriptor, which is exactly what the docs now warn about.

File: python/django_bolt/_worker_loop.py [hunk 24]
```diff
@@ -890,11 +285,10 @@ def mcp_resolve_future(fut, value):
...
 
-def run_worker_handle(loop, handle):
-    """Execute one ready Handle with the WorkerLoop installed as running."""
-    if handle.cancelled():
-        return
-    try:
-        _with_running_loop(loop, handle._run)
-    except BaseException as exc:
-        loop.call_exception_handler({"message": "Exception in WorkerLoop callback", "exception": exc, "handle": handle})
+def worker_handle_failed(loop, handle, exc):
```
[the remark this answers is not in the document]
Comment (FarhanAliRaza) [reply python/django_bolt/_worker_loop.py]: Fixed in 451f3048 (`-> None`). Note ANN isn't in the project's ruff select list, so lint was already green; added for the nit anyway.
